Feature/publish jandler and subscribe handler test

bhuvan-sapientai opened the pull request:
To enhance the detection of bug leakages and issues earlier in the release lifecycle, I have added unit tests for the following classes:

net/openhft/chronicle/queue/channel/PublishHandler.java
net/openhft/chronicle/queue/channel/SubscribeHandler.java

These are generated using the sapient.ai plugin. More: https://plugins.jetbrains.com/plugin/21709-sapient-ai-test-coder

bhuvan-sapientai committed:
Publish and Subscribe Handler Test

bhuvan-sapientai committed:
Changes to pom

tgd commented:
Hi @bhuvan-sapientai - thanks for the PR. 

We will be able to review soon.

In the meantime please sign our contributor agreement and confirm once signed - https://chronicle.software/contributor-agreement/

bhuvan-sapientai commented:
Hi @tgd I have completed the contributor agreement. Thanks for the prompt response.

tgd commented on pom.xml:
```diff
@@ -1,705 +1,658 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!--
```
Please revert these pom changes

bhuvan-sapientai replied:
Sure! On it.

tgd commented on src/test/java/net/openhft/chronicle/queue/channel/SubscribeHandlerTest.java:
```diff
@@ -0,0 +1,425 @@
+        });
+    }
+
+    //Sapient generated method id: ${37934c8d-19b2-3b9d-9f68-b07e8aedbca5}, hash: 0579A8794C3FAECD51B6D3BE6EB5408D
```
I appreciate the submission here, but please can you clean up all these dangling comments?

bhuvan-sapientai replied:
Let me remove all the comments. I'll remove all the branch conditions as well to follow the current test code style.

bhuvan-sapientai committed:
Remove unwanted comment, Add only required sapient dependencies

bhuvan-sapientai committed:
Add pom changes

bhuvan-sapientai committed:
Update pom.xml

bhuvan-sapientai commented:
Hi @tgd ! Made the expected changes. Also, I have added the dependencies that are needed around mocking and test engine. The unwanted comments are removed. Just added the branch conditions to see which block of code is under the test.

bhuvan-sapientai commented:
Hey @tgd! Can you please have a look around this PR?

bhuvan-sapientai commented:
Hey @tgd! Did you get time to look at the tests? Maybe I can help in fixing exiting issues as well. Please do let me know in case there would be any or you would need help getting around the current set of tests raised.

tgd commented:
Hi @bhuvan-sapientai - thanks for following up.

Can you please run `mvn clean install` (with Java 8) on your branch and fix up any issues. The branch needs to be able to compile/run under Java 8 and it looks like the mockito version you are pulling in is Java 11+. Some more details here: https://stackoverflow.com/questions/9170832/list-of-java-class-file-format-major-version-numbers

In addition, please can you try and remove any explicit version numbers from the `pom.xml` entries that you added - we manage version pinning via our `third-party-bom`.

The error:

```java
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:testCompile (default-testCompile) on project chronicle-queue: Compilation failure
[ERROR] /home/tomwork/code/OpenHFT/Chronicle-Queue/src/test/java/net/openhft/chronicle/queue/channel/PublishHandlerTest.java:[7,28] cannot access org.mockito.stubbing.Answer
[ERROR]   bad class file: /home/tomwork/.m2/repository/org/mockito/mockito-core/5.2.0/mockito-core-5.2.0.jar(org/mockito/stubbing/Answer.class)
[ERROR]     class file has wrong version 55.0, should be 52.0
[ERROR]     Please remove or make sure it appears in the correct subdirectory of the classpath.
```

bhuvan-sapientai committed:
Update pom.xml

bhuvan-sapientai committed:
Changes to subscribeHandlerTest

bhuvan-sapientai committed:
Merge branch 'feature/publishJandlerAndSubscribeHandlerTest' of https://github.com/bhuvan-sapientai/Chronicle-Queue into feature/publishJandlerAndSubscribeHandlerTest

bhuvan-sapientai committed:
Changes to subscribeHandlerTest

bhuvan-sapientai commented:
Hi @tgd! I ran the mvn clean install and the tests are passing now. Fixed the issues. 
![Screenshot 2024-03-27 at 10 46 43 AM](https://github.com/OpenHFT/Chronicle-Queue/assets/144414120/f6313b21-4164-4efd-8872-66c6b221a358)


I switched to java 8 and I tried removing the version. The only issue we had was in mockito and mockito-inline which uses 4.11.0 in the case of Java 8. After commenting on the version to support third-party-bom, I got the below listed issue.

```
[ERROR]   The project net.openhft:chronicle-queue:5.25ea13-SNAPSHOT (/Users/bhuvan/beta-testing/Chronicle-Queue/pom.xml) has 2 errors
[ERROR]     'dependencies.dependency.version' for org.mockito:mockito-inline:jar is missing. @ line 189, column 21
[ERROR]     'dependencies.dependency.version' for org.hamcrest:hamcrest:jar is missing. @ line 195, column 19
``` 
![Screenshot 2024-03-27 at 10 54 11 AM](https://github.com/OpenHFT/Chronicle-Queue/assets/144414120/f6fd947b-445f-49e4-b520-6da689f3f031)
![Screenshot 2024-03-27 at 10 58 48 AM](https://github.com/OpenHFT/Chronicle-Queue/assets/144414120/ee93352a-54f0-4a94-9166-bf55bb4c4118)


We would be using mockito, mockito-inline with 4.11.0, and hamcrest with 2.2 
Please let me know if there are any other changes required. I really appreciate the response.

bhuvan-sapientai committed:
Update pom.xml

tgd commented:
Hi @bhuvan-sapientai - thanks for taking the time to respond to the feedback. I have been through the code in more detail now and reviewed the AI generated tests that the Sapient plugin has produced.

I do not think these tests are useful enough for us in the current form to add into the codebase. They will need quite a bit of extension to provide useful/meaningful coverage and I don't have that time to invest right now. I am going to close this out for now. Whilst they add coverage the tests themselves are not overly clear and the comments that the AI has left in the code are not useful to the reader (particularly the branching ones). In addition, in the Queue codebase we don't use mocks much but instead prefer to use real objects (real queue objects) as much as possible, this helps us iron out more issues. I'm open to accept new tests in the future if these issues can be overcome.

bhuvan-sapientai commented:
Thanks @tgd for taking the time and providing the feedback. To understand it better, the effective coverage from the unit testing perspective of the testable methods was on the lesser side. Also, preference is for the real objects and not the mocking.
I would appreciate it if we could get on a call to understand your use case better.
